### PR TITLE
types: reduce allocation by define constant

### PR DESCRIPTION
**Pull request by ngaut**

### CODE REVIEW

File: types/field_type.go [hunk 1]
```diff
@@ -29,6 +29,23 @@ const (
 	UnspecifiedLength int = -1
 )
 
+var (
+	// Constant MySQL Tiny
+	ConstMySQLTiny = NewFieldType(mysql.TypeTiny)
```
Comment (shenli) [inline types/field_type.go]: Actually, these variables are not constants. We may meet problem when they are modified somewhere.

**Comment by shenli** [comment]
Please pass `make check`.

**Comment by ngaut** [comment]
PTAL

**Comment by breeswish** [comment]
Why are some constants UPPER CASED like `ConstMySQLLONGLONG` and some are CamelCased like `ConstMySQLTiny`?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -1364,7 +1364,7 @@ func WrapWithCastAsReal(ctx context.Context, expr Expression) Expression {
 	if expr.GetType().EvalType() == types.ETReal {
 		return expr
 	}
-	tp := types.NewFieldType(mysql.TypeDouble)
+	tp := types.ConstMySQLDouble
```
Comment (coocood) [inline expression/builtin_cast.go]: A constant must not be modified.
It's obvious if the modification near the assignment like this, but not if it's modified somewhere else.

**Pull request closed without merging**